File: app/src/main/java/io/github/ovso/blackbox/ui/base/interfaces/OnEndlessRecyclerScrollListener.java
package io.github.ovso.blackbox.ui.base.interfaces;

import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import lombok.Setter;
import lombok.experimental.Accessors;

public class OnEndlessRecyclerScrollListener extends RecyclerView.OnScrollListener {

  private int visibleThreshold = 1;
  private int lastVisibleItem, totalItemCount;
  private boolean loading;
  private LinearLayoutManager linearLayoutManager;
  private OnLoadMoreListener onLoadMoreListener;

  OnEndlessRecyclerScrollListener(OnEndlessRecyclerScrollListener.Builder builder) {
    linearLayoutManager = builder.linearLayoutManager;
    onLoadMoreListener = builder.onLoadMoreListener;
    visibleThreshold = builder.visibleThreshold;
  }

  @Override public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
    super.onScrolled(recyclerView, dx, dy);

    totalItemCount = linearLayoutManager.getItemCount();
    lastVisibleItem = linearLayoutManager.findLastVisibleItemPosition();
    if (!loading && totalItemCount <= (lastVisibleItem + visibleThreshold)) {
      loading = true;
      if (onLoadMoreListener != null) {
        onLoadMoreListener.onLoadMore();
      }
    }
  }

  public void setLoaded() {
    loading = false;
  }

  public interface OnLoadMoreListener {
    void onLoadMore();
  }

  public static class Builder implements IBuilder<OnEndlessRecyclerScrollListener> {
    @Setter @Accessors(chain = true) private int visibleThreshold = 1;
    private LinearLayoutManager linearLayoutManager;
    private OnLoadMoreListener onLoadMoreListener;

    public Builder(LinearLayoutManager $linearLayoutManager,
        OnLoadMoreListener $onLoadMoreListener) {
      linearLayoutManager = $linearLayoutManager;
      onLoadMoreListener = $onLoadMoreListener;
    }

    @Override public OnEndlessRecyclerScrollListener build() {
      return new OnEndlessRecyclerScrollListener(this);
    }
  }
}

File: app/src/main/java/io/github/ovso/blackbox/ui/base/interfaces/OnEndlessRecyclerScrollListener.kt
package io.github.ovso.blackbox.ui.base.interfaces

import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView

class OnEndlessRecyclerScrollListener internal constructor(builder: OnEndlessRecyclerScrollListener.Builder) :
    RecyclerView.OnScrollListener() {

  private var visibleThreshold = 1
  private var lastVisibleItem: Int = 0
  private var totalItemCount: Int = 0
  private var loading: Boolean = false
  private val linearLayoutManager: LinearLayoutManager?
  private val onLoadMoreListener: OnLoadMoreListener?

  init {
    linearLayoutManager = builder.linearLayoutManager
    onLoadMoreListener = builder.onLoadMoreListener
    visibleThreshold = builder.visibleThreshold
  }

  override fun onScrolled(
    recyclerView: RecyclerView,
    dx: Int,
    dy: Int
  ) {
    super.onScrolled(recyclerView, dx, dy)

    totalItemCount = linearLayoutManager!!.itemCount
    lastVisibleItem = linearLayoutManager.findLastVisibleItemPosition()
    if (!loading && totalItemCount <= lastVisibleItem + visibleThreshold) {
      loading = true
      onLoadMoreListener?.onLoadMore()
    }
  }

  fun setLoaded() {
    loading = false
  }

  interface OnLoadMoreListener {
    fun onLoadMore()
  }

  class Builder : IBuilder<OnEndlessRecyclerScrollListener> {
    var linearLayoutManager: LinearLayoutManager? = null
    var onLoadMoreListener: OnLoadMoreListener? = null
    var visibleThreshold = 1

    fun setVisibleThreshold(visibleThreshold: Int): Builder {
      this.visibleThreshold = visibleThreshold
      return this
    }

    fun setLinearLayoutManager(layoutManager: LinearLayoutManager): Builder {
      this.linearLayoutManager = layoutManager;
      return this
    }

    fun setOnLoadMoreListener(listener: OnLoadMoreListener): Builder {
      this.onLoadMoreListener = listener;
      return this
    }

    override fun build(): OnEndlessRecyclerScrollListener {
      return OnEndlessRecyclerScrollListener(this)
    }
  }
}
